### ISSUE TITLE
Swoft 2.0 - Auth and Sharing between Http and Wss Server

### ISSUE THREAD

**Issue by goddanao** [swoft-cloud/swoft#857]
QUESTION

Q | A
-- | --
Bug report? | no
Feature request? | no
Swoft version | 2.0.4-beta
Swoole version | 4.4.2-alpha
PHP version | 7.2.9
Runtime environment | Docker

First of all i'm a newby with swoft, please be patient..
I'm struggling trying to understand how can i implement a simple auth.
Looking for swoft/auth helps, but it's for swoft 1.0 and i still missing how can i share user data between http and wss and how to protect routes/services.

Can you point me in the right direction?

Thankyou very much, your work is awesome.

**Comment by stelin** [comment]
You can use middleware to achieve auth

**Comment by goddanao** [comment]
I have played with swoft/auth dev:master (which is currently migrating to swoft 2.0) and with its AuthMiddleware, implementing my custom AuthManagerInterface and AccountTypeInterface.
JWT Token are generated, stored client side, trasmitted and autheticated on each(?) Http request / Ws Message.

I got few general questions:
- authenticate in each http/ws message the token is the rigth way?
- is there a best pratice on where storing the jwt token and make it available client side? I mean, the first idea was a Http Session Cookie, but i can't understand how it works in swoft 2. LocalStorage on the browser  may work too.
- can't find a SimpleCache adapter for Redis::connection() to use as cacheClient in AuthManager, before adding external packages is it already in swoft?

While working with the websocket server i tried to implement a websocket extension for "permessage-deflate" compression support. In the WsModule@OnHandshake i can determine if the extension is needed 
```php
/**
     * @OnHandshake()
     * @param Request $request
     * @param Response $response
     * @return array [bool, $response]
     */
    public function checkHandshake(Request $request, Response $response): array
    {
        if(config('websocket.deflate')) {
            if($extensions = $request->getHeaderLine("Sec-WebSocket-Extensions")) {
                $extensions = array_flip(array_map('trim', explode(";", $extensions)));
                if(isset($extensions['permessage-deflate']))
                    $response = $response->withAddedHeader("Sec-WebSocket-Extensions", "permessage-deflate");
            }
        }
        return [true, $response];
    }
```
What i need is to switch the Parser in WsMessageContext with the DeflateParser when needed.
```php
class DeflateParser extends JsonParser implements MessageParserInterface
{
    public function encode(Message $message): string
    {
        $ctx = deflate_init(ZLIB_ENCODING_RAW);
        return deflate_add($ctx, parent::encode($message->toArray()), ZLIB_FINISH);
    }

    public function decode(string $data): Message
    {
        $ctx = inflate_init(ZLIB_ENCODING_RAW);
        return parent::decode(inflate_add($ctx, $data, ZLIB_FINISH));
    }
}
```
I'm currently implementing my custom @OnMesage handler because the default one don't send the Ext Data part of the message to the Controllers (i can't find a way to get it on the controller, but decoding again the frame->data). I think is also the right place to do the Parser swap based on handshake.

What do you think?

**Comment by inhere** [comment]
hi @goddanao 

We will release 2.0.5 soon.

The ws server's message processing method can get objects such as `Message`, `Request`, `Response`, and now can get all the information, including the `ext` part.

eg: 

```php

    /**
     * Message command is: 'test.req'
     *
     * @param Request $req
     *
     * @return void
     * @MessageMapping("req")
     */
    public function injectRequest(Request $req): void
    {
        $fd = $req->getFd();

        Session::mustGet()->push("(your FD: $fd)message data: " . \json_encode($req->getMessage()->toArray()));
    }

    /**
     * Message command is: 'test.msg'
     *
     * @param Message $msg
     *
     * @return void
     * @MessageMapping("msg")
     */
    public function injectMessage(Message $msg): void
    {
        Session::mustGet()->push('message data: ' . \json_encode($msg->toArray()));
    }
```

**Comment by inhere** [comment]
Other

- `SimpleCache` is very simple, you can use Redis to re-encapsulate it.
- `http session` and `auth` components are being written and may be released in 2.0.6

**Comment by inhere** [comment]
hi @goddanao 

The v2.0.5 has been released. please try it. Hope that has solved your problem.
